File: Chapter-05-Linked-Lists/List-Remove-Front/List-Remove-Front.py
# Given a pointer to the first node in the list, remove the head node and return the new head node

# node class
class Node(object):
    def __init__(self, value):
        self.val = value
        self.next = None

# linked list class
class lList(object):
    def __init__(self):
        self.__head = None
        self.__tail = None
        self.__length = 0

    def addFront(self, val):
        newNode = Node(val)

        if not self.__head:
            self.__head = newNode
            self.__tail = newNode
        else:
            temp = self.__head
            self.__head = newNode
            newNode.next = temp

        self.__length += 1

    def removeFront(self):
        if not self.__head:
            front = "Empty List"
        else:
            front = self.__head.val
            if self.__head == self.__tail:
                self.__head = None
                self.__tail = None
            else:
                self.__head = self.__head.next

            self.__length -= 1

        return front

    def remove(self):
        if not self.__head:
            back = "Empty List"
        else:
            current = self.__head
            back = self.__tail.val
            if current == self.__tail:
                self.__head = None
                self.__tail = None
            else:
                while current.next != self.__tail:
                    current = current.next
                current.next = None
                self.__tail = current

            self.__length -= 1

        return back

    def add(self, val):
        newNode = Node(val)

        if not self.__tail:
            self.__head = newNode
            self.__tail = newNode
        else:
            self.__tail.next = newNode
            self.__tail = newNode

        self.__length += 1

    def length(self):
        return self.__length

    def show(self):
        if not self.__head:
            print "Empty List"
        else:
            result = ''
            current = self.__head
            while current:
                result = "{}{} ".format(result, current.val)
                current = current.next

            print result

# create the list
newList = lList()
newList.show()
print newList.length()

# populate the list
for i in range(1, 11):
    newList.add(i)
newList.show()
print newList.length()

myVal = 25
# add this new value to front of the list
newList.addFront(myVal)
newList.show()
print newList.length()

# Now remove the value at the front.
print newList.removeFront()
newList.show()
print newList.length()

# remove the value at the back
print newList.remove()
newList.show()
print newList.length()

File: Chapter-05-Linked-Lists/List-Remove-Front/List-Remove-Front.js
// Given a pointer to the first node in the list, remove the head node and return the new head node

// node class
function Node(value){
  this.val = value
  this.next = null
}

// linked list class
function lList(){
  var head = null
  var tail = null
  var length = 0

  this.addFront = function(val){
    var newNode = new Node(val)
    if (!head){
      head = newNode
      tail = newNode
    } else {
      var temp = head
      head = newNode
      newNode.next = temp
    }
    length++
  }

  this.add = function(val){
    var newNode = new Node(val)
    if (!tail){
      head = newNode
    } else {
      tail.next = newNode
    }
    tail = newNode
    length++
  }

  this.removeFront = function(){
    if (!head){
      var front = "Empty List"
    } else {
      var front = head.val
      if (head == tail){
        head = null
        tail = null
      } else {
        head = head.next
      }
      length--
    }
    return front
  }

  this.remove = function(){
    if (!head){
      var back = "Empty List"
    } else {
      var back = tail.val
      var current = head

      if (current == tail){
        head = null
        tail = null
      } else {
        while (current.next != tail){
          current = current.next
        }
        tail = current
      }
      length--
    }
    return back
  }

  this.length = function(){
    return length
  }

  this.show = function(){
    if (!head){
      console.log("Empty List")
    } else {
      var result = ''
      var current = head
      while (current){
        result = result + current.val + ' '
        current = current.next
      }
      console.log(result)
    }
  }
}

// create a list first
var newList = new lList()
newList.show()
console.log(newList.length())

// populate the list
for (let i = 1; i < 11; i++){
  newList.add(i)
}
newList.show()
console.log(newList.length())

var myVal = 25
// add the new value to the front of the list
newList.addFront(myVal)
newList.show()
console.log(newList.length())

// Now remove the value at the front.
console.log(newList.removeFront())
newList.show()
console.log(newList.length())

// remove the value at the back
console.log(newList.remove())
newList.show()
console.log(newList.length())
